nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py

```diff
@@ -1156,91 +1218,130 @@ def dummy():
+        all_logits = torch.cat((ones_vect, all_logits), dim=1).long()
 
-                # TODO: do log_softmax in fp32?
-                log_probs, token_ids = torch.max(torch.nn.functional.log_softmax(output_tensor, dim=-1), dim=-1)
-                predicted_tokens_dec = torch.cat(
-                    [predicted_tokens_dec.to(token_ids.device), token_ids[:, -1].unsqueeze(1)], dim=1
-                )
-            else:
-                log_probs = torch.zeros(
-                    (predicted_tokens_dec.shape[0], predicted_tokens_dec.shape[1]), dtype=self.autocast_dtype
-                ).cuda()
-                predicted_tokens_dec = torch.zeros(
-                    (predicted_tokens_dec.shape[0], predicted_tokens_dec.shape[1] + 1),
-                    dtype=predicted_tokens_dec.dtype,
-                ).cuda()
+        # repeat source sequence beam_size times for beam search
+        _, src_length, hidden_size = enc_output.size()
+        enc_output_attn_mask = enc_output_attn_mask.repeat(1, beam_size).view(-1, src_length)
+        enc_output = enc_output.repeat(1, beam_size, 1).view(-1, src_length, hidden_size)
 
-            if self.cfg.get('pipeline_model_parallel_size', 1) > 1:
-                # Broadcast from the last pipeline stage to all other model-parallel ranks.
-                torch.distributed.broadcast(
-                    predicted_tokens_dec,
-                    parallel_state.get_pipeline_model_parallel_last_rank(),
-                    group=parallel_state.get_pipeline_model_parallel_group(),
-                )
-                torch.distributed.broadcast(
-                    log_probs,
-                    parallel_state.get_pipeline_model_parallel_last_rank(),
-                    group=parallel_state.get_pipeline_model_parallel_group(),
-                )
+        # pad_profile tracks finished hypotheses to generate only <pad> tokens
+        # if <eos> or <pad> has been generated
+        pad_profile = torch.zeros_like(scores).long()
 
-        # Reset microbatch calculator to what it was before decoding.
+        # prefixes_len tracks lengths of generated hypotheses to perform
+        # length penalty correction
+        prefixes_len = torch.zeros_like(scores).fill_(prefixes.size(1) + 1)
+
+        # reconfigure batch size for apex
+        global_batch_per_gpu = prefixes.shape[0]
+        tensor_shape[1] = prefixes.shape[0]
         _reconfigure_microbatch_calculator(
             rank=app_state.global_rank,
             rampup_batch_size=None,
             global_batch_size=global_batch_per_gpu * parallel_state.get_data_parallel_world_size(),
-            micro_batch_size=global_batch_per_gpu // num_micro_batches_before_decode,
+            micro_batch_size=global_batch_per_gpu,
             data_parallel_size=parallel_state.get_data_parallel_world_size(),
         )
-        return predicted_tokens_dec, log_probs
+
+        for _ in range(num_tokens_to_generate):
```

Comment:
the new range should be: range(num_tokens_to_generate -1) as one decode step is above, outside this loop